File: apps/web/lib/attribution.ts
/**
 * Carrying an advertisement's click identifier from the landing page to the
 * order that settles later.
 *
 * This is the part that could not be done at all before. An advertising
 * platform tags the click it sold — Google appends `gclid`, Yandex `yclid` —
 * and then wants to be told which of those clicks turned into money. Payment
 * here happens on the backend, often a day after the click and sometimes
 * through a transfer an operator confirms by hand, so the browser session that
 * carried the identifier is long gone by then. A counter script watching the
 * storefront sees the visit and never sees the sale.
 *
 * So the identifier is held, first-party, and attached to the order once the
 * order exists. It never leaves this origin except as that one attachment, and
 * this repository never sends anything to an advertising network — the operator
 * exports a file and uploads it themselves.
 *
 * Nothing here runs before consent. The capture is called from a component that
 * only mounts it after the visitor agreed, and declining clears whatever was
 * held.
 */

/** What the landing page captured. */
export type Attribution = {
  clickId?: string;
  clickSource?: string;
  source?: string;
  medium?: string;
  campaign?: string;
  content?: string;
  term?: string;
};

/**
 * The platforms that tag their own clicks, and the parameter each uses.
 *
 * A closed list. "Store whatever the URL carried" is how a session token, an
 * e-mail address, or somebody's password-reset link ends up in an analytics
 * table — the API refuses anything else anyway, and agreeing with it here means
 * the browser never holds what it could not send.
 */
const CLICK_PARAMETERS: Record<string, string> = {
  gclid: "google",
  gbraid: "google",
  wbraid: "google",
  yclid: "yandex",
  fbclid: "meta",
  msclkid: "microsoft",
  ttclid: "tiktok",
  twclid: "x",
};

/** The same shape the API validates against, so nothing is stored that it would refuse. */
const CLICK_IDENTIFIER = /^[A-Za-z0-9_.-]{6,200}$/;

const STORAGE_KEY = "omniflow.attribution";
const CHOICE_KEY = "omniflow.measurement";

/**
 * How long a click is carried.
 *
 * Thirty days is the window every major platform attributes over, and it is
 * also the answer to "how long should this sit in somebody's browser": long
 * enough that a customer who thinks about it for a week is still attributed,
 * short enough that it is not a permanent mark.
 */
const WINDOW_DAYS = 30;

type Stored = Attribution & { at: number };

/** Whether the visitor has agreed to measurement, and to this. */
export function readMeasurementChoice(): "unknown" | "granted" | "declined" {
  if (typeof window === "undefined") {
    return "unknown";
  }
  const stored = window.localStorage.getItem(CHOICE_KEY);
  return stored === "granted" || stored === "declined" ? stored : "unknown";
}

export function writeMeasurementChoice(granted: boolean) {
  if (typeof window !== "undefined") {
    window.localStorage.setItem(CHOICE_KEY, granted ? "granted" : "declined");
  }
}

/**
 * Reads the advertising parameters out of the current URL and holds them.
 *
 * Returns nothing when the visit carried none, which is the common case: most
 * visits are not from an advertisement, and storing something for every visit
 * would be a record of who arrived rather than of what an advertisement bought.
 *
 * First write wins. A customer who lands from an advertisement, browses, and
 * returns through a link somebody sent them was brought here by the
 * advertisement, and overwriting on the second visit would credit the friend.
 */
export function captureAttribution(search: string): Attribution | null {
  if (typeof window === "undefined") {
    return null;
  }
  const parameters = new URLSearchParams(search);
  const captured: Attribution = {};

  for (const [parameter, platform] of Object.entries(CLICK_PARAMETERS)) {
    const value = parameters.get(parameter)?.trim();
    if (value && CLICK_IDENTIFIER.test(value)) {
      captured.clickId = value;
      captured.clickSource = platform;
      break;
    }
  }
  captured.source = field(parameters.get("utm_source"));
  captured.medium = field(parameters.get("utm_medium"));
  captured.campaign = field(parameters.get("utm_campaign"));
  captured.content = field(parameters.get("utm_content"));
  captured.term = field(parameters.get("utm_term"));

  const anything = Object.values(captured).some((value) => value);
  if (!anything) {
    return readAttribution();
  }
  const existing = readAttribution();
  if (existing) {
    return existing;
  }
  const stored: Stored = { ...prune(captured), at: Date.now() };
  window.localStorage.setItem(STORAGE_KEY, JSON.stringify(stored));
  return prune(captured);
}

/** What is being carried, or null once it has aged out. */
export function readAttribution(): Attribution | null {
  if (typeof window === "undefined") {
    return null;
  }
  const raw = window.localStorage.getItem(STORAGE_KEY);
  if (!raw) {
    return null;
  }
  try {
    const stored = JSON.parse(raw) as Stored;
    if (!stored.at || Date.now() - stored.at > WINDOW_DAYS * 86_400_000) {
      window.localStorage.removeItem(STORAGE_KEY);
      return null;
    }
    const { at: _ignored, ...attribution } = stored;
    return Object.values(attribution).some((value) => value) ? attribution : null;
  } catch {
    window.localStorage.removeItem(STORAGE_KEY);
    return null;
  }
}

/** Forgets it. Called when somebody declines measurement, and after it is attached. */
export function clearAttribution() {
  if (typeof window !== "undefined") {
    window.localStorage.removeItem(STORAGE_KEY);
  }
}

/**
 * Bounds a campaign value the same way the API does, so the browser never holds
 * something the attachment would be refused for.
 */
function field(value: string | null): string | undefined {
  const trimmed = value?.trim();
  if (!trimmed) {
    return undefined;
  }
  // The escapes are written out rather than the characters themselves, so this
  // source file stays free of the control bytes it is stripping.
  // biome-ignore lint/suspicious/noControlCharactersInRegex: removing them is the point.
  return trimmed.replace(/[\u0000-\u001f\u007f]/g, "").slice(0, 120) || undefined;
}

function prune(attribution: Attribution): Attribution {
  return Object.fromEntries(
    Object.entries(attribution).filter(([, value]) => Boolean(value)),
  ) as Attribution;
}

File: internal/adtracking/attribution.go
package adtracking

import (
	"errors"
	"fmt"
	"net/url"
	"regexp"
	"strings"
)

// The click identifier, and why it has to be carried rather than observed.
//
// An advertising platform tags the click it sold: Google appends `gclid`,
// Yandex `yclid`, Meta `fbclid`, Microsoft `msclkid`, TikTok `ttclid`. The
// platform then wants to be told which of those clicks turned into a paid
// order, so it can attribute the money and optimise the bidding.
//
// Nothing in a normal analytics setup can tell it. Payment here happens on the
// backend, often a day after the click, sometimes through a bank transfer an
// operator confirms by hand, and always outside the browser session that
// carried the identifier. A counter script watching the storefront sees the
// visit and never sees the money. That is why "no advertising channel can be
// attributed at all" was true even for an installation that had pasted a
// counter in.
//
// So the identifier is carried: captured on the first visit, held first-party,
// attached to the order, and exported when that order settles. What the
// operator uploads to the platform is a file — this repository makes no request
// to any advertising network, with or without consent. Sending a customer's
// purchase to Google because a settings toggle was on is not a thing an
// operator can be asked to discover from behaviour.

// Attribution is one order's advertising origin.
//
// It hangs off the order rather than the customer, deliberately. An order is a
// conversion and has an amount and a date, which is exactly what an offline
// upload needs. A customer is a person, and a click identifier attached to a
// person for the life of their account is a profile — a different thing, with
// different obligations, that nothing here needs.
type Attribution struct {
	// ClickID is the platform's own identifier for the click it sold.
	ClickID string `json:"clickId,omitempty"`
	// ClickSource names which platform issued it, so an export can be filtered
	// per network without parsing the identifier's shape.
	ClickSource string `json:"clickSource,omitempty"`

	Source   string `json:"source,omitempty"`
	Medium   string `json:"medium,omitempty"`
	Campaign string `json:"campaign,omitempty"`
	Content  string `json:"content,omitempty"`
	Term     string `json:"term,omitempty"`
}

// clickParameters maps a query parameter to the platform that issues it.
//
// The list is closed. An unrecognised parameter is not captured, because "store
// whatever the URL carried" is how a session identifier, an e-mail address, or
// somebody's password reset token ends up in an analytics table.
var clickParameters = map[string]string{
	"gclid":   "google",
	"gbraid":  "google",
	"wbraid":  "google",
	"yclid":   "yandex",
	"fbclid":  "meta",
	"msclkid": "microsoft",
	"ttclid":  "tiktok",
	"twclid":  "x",
}

// clickIdentifier is what every platform's identifier looks like: an opaque
// token. Anchored and bounded, because it is stored and later written into a
// CSV an operator uploads.
var clickIdentifier = regexp.MustCompile(`^[A-Za-z0-9_.-]{6,200}$`)

// campaignField bounds a UTM value. They are operator-authored strings that
// travel through a customer's URL bar, so they are length-bounded and stripped
// of anything that is not printable — a campaign name is a label, not a payload.
const maxCampaignField = 120

var (
	ErrNoAttribution      = errors.New("no advertising parameters")
	ErrMalformedClickID   = errors.New("malformed click identifier")
	ErrUnknownClickSource = errors.New("unknown click source")
)

// Capture reads the advertising parameters out of a landing URL's query.
//
// It returns ErrNoAttribution when there is nothing to record, which is the
// common case: most visits are not from an advertisement, and an attribution
// row for every visit would be a log of who arrived rather than a record of
// what an advertisement bought.
func Capture(rawQuery string) (Attribution, error) {
	values, err := url.ParseQuery(rawQuery)
	if err != nil {
		return Attribution{}, ErrNoAttribution
	}
	return CaptureValues(values)
}

// CaptureValues is Capture over already-parsed values.
func CaptureValues(values url.Values) (Attribution, error) {
	var attribution Attribution
	for parameter, platform := range clickParameters {
		candidate := strings.TrimSpace(values.Get(parameter))
		if candidate == "" {
			continue
		}
		if !clickIdentifier.MatchString(candidate) {
			// A malformed identifier is dropped rather than refused. The visitor
			// did nothing wrong and is trying to buy something; failing their
			// landing page over an advertising parameter would be a worse
			// outcome than not attributing one click.
			continue
		}
		attribution.ClickID = candidate
		attribution.ClickSource = platform
		break
	}

	attribution.Source = campaignField(values.Get("utm_source"))
	attribution.Medium = campaignField(values.Get("utm_medium"))
	attribution.Campaign = campaignField(values.Get("utm_campaign"))
	attribution.Content = campaignField(values.Get("utm_content"))
	attribution.Term = campaignField(values.Get("utm_term"))

	if attribution.Empty() {
		return Attribution{}, ErrNoAttribution
	}
	return attribution, nil
}

// CheckAttribution validates an attribution that arrived from a browser rather
// than being captured here, which is what the account endpoint receives.
func CheckAttribution(attribution Attribution) error {
	if attribution.ClickID != "" {
		if !clickIdentifier.MatchString(attribution.ClickID) {
			return ErrMalformedClickID
		}
		if _, known := knownSources()[attribution.ClickSource]; !known {
			return fmt.Errorf("%w: %q", ErrUnknownClickSource, attribution.ClickSource)
		}
	}
	if attribution.Empty() {
		return ErrNoAttribution
	}
	return nil
}

// Clean bounds the campaign fields of an attribution that arrived from a
// browser, so a caller cannot store two kilobytes of anything under `utm_term`.
func Clean(attribution Attribution) Attribution {
	return Attribution{
		ClickID:     strings.TrimSpace(attribution.ClickID),
		ClickSource: strings.TrimSpace(attribution.ClickSource),
		Source:      campaignField(attribution.Source),
		Medium:      campaignField(attribution.Medium),
		Campaign:    campaignField(attribution.Campaign),
		Content:     campaignField(attribution.Content),
		Term:        campaignField(attribution.Term),
	}
}

// Empty reports whether there is nothing worth storing.
func (attribution Attribution) Empty() bool {
	return attribution.ClickID == "" && attribution.Source == "" &&
		attribution.Medium == "" && attribution.Campaign == "" &&
		attribution.Content == "" && attribution.Term == ""
}

func knownSources() map[string]bool {
	sources := map[string]bool{}
	for _, platform := range clickParameters {
		sources[platform] = true
	}
	return sources
}

// campaignField trims, bounds, and strips control characters.
//
// The control-character strip is what keeps a newline out of a value that later
// becomes a CSV cell an operator uploads to an advertising platform.
func campaignField(value string) string {
	value = strings.TrimSpace(value)
	if value == "" {
		return ""
	}
	cleaned := strings.Map(func(symbol rune) rune {
		if symbol < 0x20 || symbol == 0x7f {
			return -1
		}
		return symbol
	}, value)
	runes := []rune(cleaned)
	if len(runes) > maxCampaignField {
		runes = runes[:maxCampaignField]
	}
	return strings.TrimSpace(string(runes))
}
